src/Xml/Dom/Locator/Node/value.php

```diff
@@ -17,7 +17,10 @@
  *
  * @throws CoercionException
  */
-function value(DOMNode $node, TypeInterface $type)
+function value(\DOM\Node $node, TypeInterface $type)
 {
-    return $type->coerce($node->nodeValue);
+    // TODO : nodeValue did entity substitution
```

Comment:
TODO

![Screenshot 2024-02-23 at 11 20 31](https://github.com/veewee/xml/assets/1618158/5632d12c-2818-43c5-a84f-914dd144b3a2)

It's not possible to use nodeValue anymore on elements.


Although:

> However, it can return a string too for elements for example.

This doesn't seem to always work and is not very transparent.


It's probably an edge case anyways - but I'm not sure how to continue yet.

For example:

In this new implementation I can set XML with substituted entities. 
Whilst reading though, the entities are not substituted back and there seems to be no way to get the substituted value.
The elements nodeValue is also empty:

```php
$doc = \DOM\XMLDocument::createEmpty();
$el = $doc->createElement('el');
$text1 =$doc->createTextNode('<wor " ld>');
$text2 =$doc->createTextNode(htmlspecialchars('<wor " ld>', ENT_XML1|ENT_QUOTES));
//$fragment->appendXML(htmlspecialchars('<wor " ld>', ENT_XML1|ENT_QUOTES));
$doc->append($el);
$el->append($text1);
$el->append($text2);

echo $doc->saveXML().PHP_EOL;
echo $el->textContent.PHP_EOL;
echo $el->nodeValue.PHP_EOL;
echo $text1->nodeValue.PHP_EOL;
echo $text2->nodeValue.PHP_EOL;
```

```
<?xml version="1.0" encoding="UTF-8"?>
<el>&lt;wor " ld&gt;&amp;lt;wor &amp;quot; ld&amp;gt;</el>
<wor " ld>&lt;wor &quot; ld&gt;

<wor " ld>
&lt;wor &quot; ld&gt;
```

So in this case, there is no way to unsubstitute: `&lt;wor &quot; ld&gt;` (maybe for the good, who knows... )


In legacy API, this just worked:

```php
$doc = new \DOMDocument();
$el = $doc->createElement('el');
$el->nodeValue = '<wor " ld>' . htmlspecialchars('<wor " ld>', ENT_XML1|ENT_QUOTES);

$doc->append($el);

echo $doc->saveXML().PHP_EOL;
echo $el->textContent.PHP_EOL;
echo $el->nodeValue.PHP_EOL;
```

Both writing and reading from nodeValue was smart enough to do the trick.


```
<?xml version="1.0"?>
<el>&lt;wor " ld&gt;&lt;wor " ld&gt;</el>

<wor " ld><wor " ld>
<wor " ld><wor " ld>
```

Reply:
> Although:
>     However, it can return a string too for elements for example.
> This doesn't seem to always work and is not very transparent.

This was directed at old DOM. nodeValue always returns NULL for new DOM as spec demands.

There's no way beyond appendXML to unsubstitute entities right now, which is not ideal because it will also interpret tags. Supporting unsubstituting would require a custom API.

Or right now you could work around it by creating an element in legacy DOM, and importing it via `importLegacyNode`.

Reply:
I would like to avoid adding fallbacks to legacy behaviour, cause that's probably not the way to go.

Not having a way to read / write substituted content is a bit annoying from my point of view. For the given example, it might be solvable, but it comes even more complex if the document specifies its own XML entities?
Then the consumer somehow needs to map the content against the provided list of entities?

About `appendXML`: I've tried that first but it seems to parse the string as it were XML, making it not very usefull for these specific functions. The TextNode seems to be doing a good job in sanitizing input and feels like the way to go here. It already covers entity substitution to some extend, but not as "good" (or bad) as the legacy system.

Reply:
Then I think the proper fix would be to introduce a custom API to allow the substitution behaviour.
For example, this could be a field on the DOM\Text class, and/or a variant of the createTextNode. Perhaps the latter is even sufficient because you could always use replaceNode.
We also need a name for that method then. What do you think?

Reply:
The thing is that an element can consist out of multiple `Text`(and other) nodes.
So it might make sense to have something like `readSubstitutedValue()` and `writeSubstitutedValue($rawValue)` at both `Element`and `Text`level? That makes the intention clear and maybe even lets the end users think about "do I really want to do this less secure operation?"

Reply:
That works for me, but I'd propose to use get/set instead of write/read for consistency with other method names.
so that would be either:
```
function getSubstitutedValue(): string;
function setSubstitutedValue(string $value): void;
```

or as a field: `public string $substitutedValue;`
Probably the latter is the most convenient, but I wonder whether this will communicate clearly this can be insecure.

Reply:
Whatever works best for you is fine. The insecurity can always be documentated and explained.